ml/src/test/scala/ai/h2o/sparkling/ml/algos/H2OKMeansTest.scala

```diff
@@ -0,0 +1,121 @@
+import org.scalatest.{FunSuite, Matchers}
+import water.api.TestUtils
+import water.exceptions.H2OModelBuilderIllegalArgumentException
+
+@RunWith(classOf[JUnitRunner])
+class H2OKMeansTest extends FunSuite with Matchers with SharedH2OTestContext {
+
+  override def createSparkContext = new SparkContext("local[1]", this.getClass.getSimpleName, conf = defaultSparkConf)
+
+  private lazy val dataset = spark.read
+    .option("header", "true")
+    .option("inferSchema", "true")
+    .csv(TestUtils.locate("smalldata/iris/iris_wheader.csv"))
+
+  test("H2OKMeans Pipeline serialization and deserialization") {
+    val algo = new H2OKMeans()
+      .setSplitRatio(0.8)
+      .setSeed(1)
+      .setK(3)
+      .setFeaturesCols("sepal_len", "sepal_wid", "petal_len", "petal_wid")
+
+    val pipeline = new Pipeline().setStages(Array(algo))
+    pipeline.write.overwrite().save("ml/build/kmeans_pipeline")
+    val loadedPipeline = Pipeline.load("ml/build/kmeans_pipeline")
+    val model = loadedPipeline.fit(dataset)
+
+    model.write.overwrite().save("ml/build/kmeans_pipeline_model")
+    val loadedModel = PipelineModel.load("ml/build/kmeans_pipeline_model")
+
+    loadedModel.transform(dataset).count()
+  }
+
+  test("H2OKMeans' result is directly in prediction column") {
+    val algo = new H2OKMeans()
+      .setSplitRatio(0.8)
+      .setSeed(1)
+      .setK(3)
+      .setFeaturesCols("sepal_len", "sepal_wid", "petal_len", "petal_wid")
+
+    val model = algo.fit(dataset)
+    val transformed = model.transform(dataset)
+    assert(transformed.select("prediction").head() == Row(0))
+    assert(transformed.select("prediction").distinct().count() == 3)
+  }
+
+  test("H2OKMeans' full result in prediction details column") {
+    val algo = new H2OKMeans()
+      .setSplitRatio(0.8)
+      .setSeed(1)
+      .setK(3)
+      .setWithDetailedPredictionCol(true)
+      .setFeaturesCols("sepal_len", "sepal_wid", "petal_len", "petal_wid")
+
+    val model = algo.fit(dataset)
+    val transformed = model.transform(dataset)
+    assert(transformed.select("detailed_prediction.cluster").head().getInt(0) == 0)
+    assert(transformed.select("detailed_prediction.distances").head().getAs[Seq[Double]](0).length == 3)
+  }
+
+  test("H2OKMeans user points") {
+    val algo = new H2OKMeans()
+      .setSplitRatio(0.8)
+      .setSeed(1)
+      .setK(3)
+      .setUserPoints(Array(Array(4.9, 3.0, 1.4, 0.2), Array(5.6, 2.5, 3.9, 1.1), Array(6.5, 3.0, 5.2, 2.0)))
+      .setFeaturesCols("sepal_len", "sepal_wid", "petal_len", "petal_wid")
+
+    val model = algo.fit(dataset)
+    assert(model.transform(dataset).select("prediction").head() == Row(0))
+  }
+
+  test("H2OKMeans single string column input without converting to categorical") {
+    val algo = new H2OKMeans()
+      .setSplitRatio(0.8)
+      .setSeed(1)
+      .setK(3)
+      .setFeaturesCols("class")
+      .setAllStringColumnsToCategorical(false)
+
+    intercept[H2OModelBuilderIllegalArgumentException] {
```

Comment:
Does the exception message contain a reasonable message for SW users?

Reply:
We can do better on our side, will catch and re-throw, thanks